web-server-lib/web-server/http/redirect.rkt

```diff
@@ -8,8 +8,14 @@
 ; redirection-status = (make-redirection-status nat bytes)
 (define-struct redirection-status (code message))
 
-(define permanently (make-redirection-status 301 #"Moved Permanently"))
-(define temporarily (make-redirection-status 302 #"Moved Temporarily"))
+(define permanently
+  ;; NOTE: 308 permanent redirect is not supported by
+  ;; Internet Explorer on Windows 7 or 8.1 as of 2019-02-26.
+  ;; (IE on Windows 10 does support it.)
+  ;; https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/308
+  ;; https://tools.ietf.org/html/rfc7538#section-4
+  (make-redirection-status 301 #"Moved Permanently"))
+(define temporarily (make-redirection-status 307 #"Temporary Redirect"))
```

Comment:
Isn't this technically a breaking change?

If I have a request handler that receives a `POST` request and I am using `redirect-to`, after this change, the browser will attempt to make a `POST` request to the page I am redirecting to, which may or may not support `POST` requests.

Pseudocode example of what I mean:


```racket
(define-values (dispatch _)
  (dispatch-rules
    [("login") #:method (or "get" "post") login-page]
    [("login-success") login-success-page]))

(define (login-page req)
  (cond
     [(and posted? valid-login?)
        (redirect-to "/login-success")]))

(define (login-sucess req)
  (response/xexpr '(h1 "Success")))
```

I believe this code would break after this change. 

IMO, this module should contain and expose the "new" redirection codes, but the default behaviour should continue being a 302.

Reply:
I think there is a pragmatic question of whether existing programs might be relying on the behavior you illustrate.

On principle, my understanding is that the above code is relying on user-agent-defined behavior that is either not guaranteed (in HTTP 1.1, per [RFC 7231 §6.4.3](https://tools.ietf.org/html/rfc7231#section-6.4.3) or is explicitly called "erroneous" (in HTTP 1.0, per [RFC 1945 §9.3](https://tools.ietf.org/html/rfc1945#section-9.3). At the Racket level, `redirect-to` has not been documented as using any particular HTTP status code for the redirection.

On the pragmatic question, I always use `redirect/get` for the post-redirect-get pattern, so I assumed others did, too. If we think many people are using `redirect-to` with `temporarily` (explicitly or implicitly) and expecting it to work for post-redirect-get (or, at least, to work as well as it does now in most common user-agents), then changing the meaning of `temporarily` is probably wrong. In that case, I could see adding either two more statuses like `temporarily/same-method` or adding an optional keyword argument like `#:keep-method?` (but with a better name, so the default would apply to `301`, `302`, and `303`.

I don't see any reason for `redirect/get` to be using `302` instead of `303 See Other`, regardless.

Reply:
@Bogdanp Can you check to see if it does break?

@LiberalArtist Can you do a quick search on github for uses of `redirect-to` that might be at risk? I checked in the Racket code I have access to and did not find any.

Reply:
FWIW, I do have (non-open source) code that does this exact thing. I realize `redirect/get` is better under many circumstances, but the UX of continuations is lacking when it comes to URLs that ought to be shareable between users, so I think this behavior would be missed for those types of situations.

@jeapostrophe it does break. Here's a minimal working example: https://gist.github.com/Bogdanp/4fce05956eba3838bd9a7aed3be48f83 . Change the 307 into a 302 and everything works as expected.

@LiberalArtist this is a fairly common pattern I've seen in languages other than Racket and  one I've used myself instead of relying on continuations in cases where the user must be able to easily share application URLs with others. All of the major browsers behave the same way with regards to HTTP 302 as far as I am aware. They are consistent with each other even though they may not necessarily be consistent with the spec.

> On the pragmatic question, I always use `redirect/get` for the post-redirect-get pattern, so I assumed others did, too. If we think many people are using `redirect-to` with `temporarily` (explicitly or implicitly) and expecting it to work for post-redirect-get (or, at least, to work as well as it does now in most common user-agents), then changing the meaning of `temporarily` is probably wrong. In that case, I could see adding either two more statuses like `temporarily/same-method` or adding an optional keyword argument like `#:keep-method?` (but with a better name, so the default would apply to `301`, `302`, and `303`.

My personal preference would be to add additional statuses and let the user choose which one is returned in the response. 

> I don't see any reason for `redirect/get` to be using `302` instead of `303 See Other`, regardless.

It's possible that clients that only support HTTP/1.0 won't know how to handle HTTP 303 responses. I don't mind either way, but is something to consider.